python/src/equistore/operations/slice.py

```diff
@@ -0,0 +1,395 @@
+    block upon slicing is reduced to nothing, i.e. in the case that it has none
+    of the specified `samples_to_slice` or `properties_to_slice`, an empty block
+    will be returned but will still be accessible by its key. User warnings will
+    be issued if any blocks are sliced to contain no values.
+
+    For the empty blocks that may be returned, although there will be no actual
+    values in its TensorBlock.values tensor, the shape of this tensor will be
+    non-zero in the dimensions that haven't been sliced. This is created by
+    slicing the input TensorBlock, as opposed to just returning an
+    artificially-created empty one (with no shape or dimensions), and is
+    intentional. It allows the slicing of dimensions to be tracked.
+
+    For example, if a TensorBlock of shape (52, 1, 5) is passed, and only some
+    samples are specified to be sliced but none of these appear in the input
+    TesnorBlock, the returned TensorBlock values array will be empty, but its
+    shape will be (0, 1, 5) - i.e. the samples dimension has been sliced to zero
+    but the components and properties dimensions remain in-tact. The same logic
+    applies to any Gradient TensorBlocks the input TensorBlock may have
+    associated with it.
+
+    See the documentation for the `slice_block()` to see how an individual
+    :py:class:`TensorBlock` is sliced.
+
+    :param tensor: the input :py:class:`TensorMap` to be sliced.
+    :param samples_to_slice: a :py:class:`Labels` object containing the names
+        and indices of samples to keep in the sliced :py:class:`TensorBlock`
+        output, or each of the sliced :py:class:`TensorBlock` objects of the
+        output :py:class:`TensorMap`. Default value of None indicates no slicing
+        along the samples dimension should occur.
+    :param properties_to_slice: a :py:class:`Labels` object containing the names
+        and indices of properties to keep in the sliced :py:class:`TensorBlock`
+        output, or each of the sliced :py:class:`TensorBlock` objects of the
+        output :py:class:`TensorMap`. Default value of None indicates no slicing
+        along the properties dimension should occur.
+
+    :return: a :py:class:`TensorMap` that corresponds to the sliced input tensor.
+    """
+
+    # Perform input checks
+    assert isinstance(tensormap, TensorMap), (
+        "Input tensor must be an equistore `TensorMap` object."
+        + " If attempting to slice an equistore `TensorBlock`,"
+        + " use the `slice_block()` function instead."
+    )
+    if samples_to_slice is None and properties_to_slice is None:
+        raise ValueError(
+            "Must specify either some samples or properties (or both) to slice by."
+        )
+    if samples_to_slice is not None:
+        assert isinstance(
+            samples_to_slice, Labels
+        ), "samples_to_slice must be passed as a equistore `Labels` object."
```

Comment:
This would be better as a TypeError (users can disable asserts, and the error message is clearer with a TypeError)